vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,330 @@
+# Score-Based Generative Modeling
+#
+# Score-Based Generative Modeling through Stochastic Differential Equations.
+# Yang Song, Jascha Sohl-Dickstein, Diederik P. Kingma, 
+# Abhishek Kumar, Stefano Ermon, and Ben Poole.
+# https://arxiv.org/pdf/2011.13456.pdf
+
+using MLDatasets
+using Flux
+using Flux: @functor, chunk, params
+using Flux.Data: DataLoader
+using Parameters: @with_kw
+using BSON
+using CUDA
+using Images
+using Logging: with_logger
+using ProgressMeter: Progress, next!
+using TensorBoardLogger: TBLogger, tb_overwrite
+using Random
+
+"""
+Projection of Gaussian Noise onto a time vector.
+
+# Notes
+This layer will help embed our random times onto the frequency domain. \n
+W is not trainable and is sampled once upon construction - see assertions below.
+
+# References
+paper-  https://arxiv.org/abs/2006.10739 \n
+layers- https://fluxml.ai/Flux.jl/stable/models/basics/#Building-Layers
+"""
+function GaussianFourierProjection(embed_dim, scale)
+    # Instantiate W once
+    W = randn(Float32, embed_dim ÷ 2) .* scale
+    # Return a function that always references the same W
+    function GaussFourierProject(t)
+        t_proj = t' .* W * Float32(2π)
+        [sin.(t_proj); cos.(t_proj)]
+    end
+end
+
+"""
+Helper function that computes the *standard deviation* of 𝒫₀ₜ(𝘹(𝘵)|𝘹(0)).
+
+# Notes
+Derived from the Stochastic Differential Equation (SDE):    \n
+                𝘥𝘹 = σᵗ𝘥𝘸,      𝘵 ∈ [0, 1]                   \n
+
+We use properties of SDEs to analytically solve for the stddev
+at time t conditioned on the data distribution. \n
+
+We will be using this all over the codebase for computing our model's loss,
+scaling our network output, and even sampling new images!
+"""
+function marginal_prob_std(t, sigma=25.0f0)
+    sqrt.((sigma .^ (2t) .- 1.0f0) ./ 2.0f0 ./ log(sigma))
+end
```

Comment:
This could probably be a one-liner. More generally, I'm surprised there isn't a pre-made function for this in some existing library (or is there?)

Reply:
Not to my knowledge... do one liner functions still have docstrings? I feel like some level of explanation is needed here to motivate the operation. I could also move as a normal comment if one-liner is much preferred

Reply:
<img width="886" alt="image" src="https://user-images.githubusercontent.com/43712099/164579943-074e9a3f-9bc2-402a-8bb0-f070593dc47c.png">

Reply:
Yes, docstrings work exactly the same way.